File: plugins/collector/event/apps/software/php.go
package software

import (
	"collector/cache/process"
	"collector/event/apps"
	"errors"
	"regexp"
	"strings"
)

type PHP struct {
	version string
	r       *regexp.Regexp
}

func (PHP) Name() string { return "php" }

func (PHP) Type() string { return "software" }

func (p PHP) Version() string { return p.version }

// only fpm
func (PHP) Match(p *process.Process) bool { return p.Name == "php-fpm" }

func (p *PHP) Run(proc *process.Process) (m map[string]string, err error) {
	if p.r == nil {
		p.r = regexp.MustCompile(`PHP\s(\d+\.)+\d+`)
	}
	result, err := apps.Execute(proc, "-v")
	if err != nil {
		return nil, err
	}
	str := p.r.FindString(result)
	if str == "" {
		err = errors.New("version not found")
		return
	}
	p.version = strings.TrimPrefix(str, "PHP ")
	return
}

func init() {
	apps.Regist(&PHP{})
}

File: plugins/collector/src/event/apps/software/php.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Php {
    version: String,
}

impl Php {
    pub fn new() -> Self { Php { version: String::new() } }
}

impl IApp for Php {
    fn name(&self)     -> &'static str { "php" }
    fn app_type(&self) -> &'static str { "software" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "php-fpm" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["-v"])?;
        // "PHP 8.1.13 (fpm-fcgi) …"
        self.version = find_version(&out);
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
